### DISCUSSION TITLE
Using Git (and Gix) as an object storage

### DISCUSSION THREAD

**Discussion by 35VLG84** [GitoxideLabs/gitoxide#1723] [General]
Hello Gitoxide project,

I have ported a git based application from Scala to Rust, and Gix has played a major role in making that possible! 

[Tackler-ng](https://github.com/e257-fi/tackler-ng)  is bookkeeping engine and tool with native GIT SCM support for plain text accounting.  It can read transaction data as text files either from the disk, or directly from bare git repository. 

Part of the test suite are tests with large amount of transactions, up to 1 million transaction (files).  In these tests git repository access (by Gix)  is about **20% faster** than reading from plain disk!  

The first priority has been to reach feature parity with the old code base so this is totally un-optimized code, but the code to access repository is here (if it's helpful to somebody):

https://github.com/e257-fi/tackler-ng/blob/0fb1ee326338cf35ea8e7994b8d228766ce4724d/tackler-core/src/parser/tackler_txns.rs#L59

Thank you for making Gitoxide (and rusty tackler possible) - it has been fun journey (starting from Gix 0.35.0)!

**Comment by Byron** [comment]
Thanks so much for sharing, I am very excited about `Tackler-ng` and impressed by the amount of features *and* its presentation! It's something I'd want to try myself as well!

> In these tests git repository access (by Gix) is about **20% faster** than reading from plain disk

If it's Git vs a million files, then I think Git could be much faster. If it's Git vs a couple of bigger files, then I think the 20% performance advantage are due to compression.

From what I can tell, Git is used for snapshotting and each change becomes a commit. For reading values, `HEAD^{tree}` is  used as if it was a filesystem.

When looking at the linked code, I thought that maybe one would even want to support `revspecs`, and if you started with 0.35, chances are `repo.revparse_single()` wasn't available yet.

When opening the repository, and it's from a well-known location, I'd recommend using `gix::open()` instead so it won't traverse upward and maybe find something unexpected.

Otherwise I think it's pretty optimal, but depending on how long parsing transaction files takes, one might consider parallelising that.

That's all my thoughts on the Git part :)!

**Reply by 35VLG84** [reply]
Thank you so much for the comments!  I will include those into the future changes.

> From what I can tell, Git is used for snapshotting and each change becomes a commit. For reading values, HEAD^{tree} is used as if it was a filesystem.

Yes, this exactly the idea. It makes it possible to switch between working copy and git at the same time during data entry, and generating reports so that the state of accounting journal is always well defined ([Report Audit example](https://tackler.e257.fi/docs/quickstart/#lets_play_for_real_git_storage_and_audit_mode) and [Journal setup Example](https://tackler.e257.fi/docs/configuration/examples/))


> Otherwise I think it's pretty optimal, but depending on how long parsing transaction files takes, one might consider parallelising that.

Thanks, that's really good to hear. The parsing in general and the current parser (ANTLR and Antlr4Rust) are quite slow, so parallelising is definitely something to change in the future.  The old scala code was parallel, and it was about same speed with 4 cores as the current rust implementation is with single thread. And let's not talk about anything about memory usage (scala vs. rust) ;)  I am also planning to replace the parser with winnow, as Gix is using it and that way the amount of dependencies will be little bit smaller.

Thanks again for the comments, those are highly appreciated!

**Reply by Byron** [reply]
20 minutes into studying the manual I think I got a better idea what Tackler is, and it's no less enticing now! I can totally imagine how it can be used to automate business finances entirely, and even how it could be distributed among multiple locations (Git with merge-strategy union maybe, + Sharding :)?).

And now that it's written in Rust it really just has to move onto my list of projects to track.

Please do feel free to ask or ping me for reviews of the Git related parts, happy to help.

PS: Something I am still missing is a better idea of how I could 'move' items/money from one account to another, and how it assures that the entered data makes sense maybe.

**Reply by 35VLG84** [reply]
> imagine how it can be used to automate business finances entirely, and even how it could be distributed among multiple locations (Git with merge-strategy union maybe, + Sharding :)?).

Exactly! There are multiple aspects on that:

- It's distributed by nature, which makes merging trivial with correct journal sharding
- All parties can have well defined state of journal (e.g. branch name or commit id)
- Accounting Reports have explicit reference what was the state of journal when the report was generated
- It's trivial to apply "two-pairs of eyes" -principle - it's just a PR with approval workflow
  - The transactions are human readable, so it's easy to train people to review the PR
- Tackler is modeled as processor for  stream of  immutable transactions, where the processing order of transactions doesn't  matter
   - It's also possible to remove transactions from the transaction set (aka filter them), it of course affects the end results but it doesn't invalidate processing. This makes it possible to filter transactions and to do partial processing of transactions sets.
 
About how items / money move from one account to another

The first principle is that there must be at least two accounts with value (postings) in the transaction, and the sum of postings must be zero. This is the same principle as with normal  double-entry bookkeeping with debit and credit accounts, but with plain text accounting system that's denoted by sign (+/-) for accounts. 

Here are two examples about how items or money could move around:

* [How to track trimix at gas filling station](https://tackler.e257.fi/docs/examples/trimix-filling-station/)
* [How to find transactions which moves money from Account A to B](https://tackler.e257.fi/docs/examples/filter-for-two-accounts/)

> Please do feel free to ask or ping me for reviews of the Git related parts, happy to help.

That's precious offering! I will keep it in mind when I'm a bit of a pickle with git!

**Reply by Byron** [reply]
Thank you! I found both examples very interesting!

With the one about filter definition I wondered if it could just read the filter from a file.

And one more thought I had: What if there are several partitions, like multiple stores of a chain, and these should operate locally but submit their transactions at the end of each day. They shouldn't, however, have full access to all the companies transactions and accounts.

This would mean there would have to be 'links' to accounts they don't have, has they would still have to denote transactions from or to accounts they don't own. For them, these would be more like proxies as they only have transactions that they are involved in.

This isn't a feature request :D, just a thought.

**Reply by 35VLG84** [reply]
Hi, the idea of file based filter definition is good, I have to think if there would be feasible way to have it with variables (e.g. for timestamps).

For the second one, that's really interesting question and thought, and my first impression was that it's not possible, and would be quite tricky to implement. But there is no need to file a feature request ;)  

This will be quite long, but there is more git than tackler on this one - the spoiler is `git merge --allow-unrelated-histories `

Establish Acme Inc. Headquarters
```bash
cd tackler-ng
mkdir -p tmp/acme-inc
cd tmp/acme-inc
git init --bare acme-hq.git
git clone acme-hq.git/
cd acme-hq
git commit -m init --allow-empty
git push
cd ..
```

Acme Inc. Store A
```bash
# inside 'tackler-ng/tmp/acme-inc' 
git init --bare acme-store-A.git
git clone acme-store-A.git/
cd acme-store-A
git commit -m init --allow-empty
git push
cd ..
```
Acme Inc Store B
```bash
# inside 'tackler-ng/tmp/acme-inc' 
git init --bare acme-store-B.git/
git clone  acme-store-B.git/
cd acme-store-B
git commit -m init --allow-empty
git push
cd ..
```
Sales event on Store A!
```bash
# inside 'tackler-ng/tmp/acme-inc' 
cd acme-store-A
mkdir -p txns/store-a
cat << EOF >> txns/store-a/txns-001.txn
2024-12-17 'store a txn
   Assets:Cashier  42 €
   Warehouse:Acmerator  -1 Unit @ 42 €
EOF
git add .
git commit -m 'store-a'
git push
cd ..
```
Sales event on Store B!
```bash
# inside 'tackler-ng/tmp/acme-inc' 
cd acme-store-B
mkdir -p txns/store-b
cp ../acme-store-A/txns/store-a/txns-001.txn txns/store-b/
sed -i 's/store a/store b/' txns/store-b/txns-001.txn
git add .
git commit -m 'store b'
git push
cd ..
```
Busy day at Accounting department of Acme Inc Headquarters
```bash
# inside 'tackler-ng/tmp/acme-inc' 
cd acme-hq
git remote add store-a ../acme-store-A.git
git remote add store-b ../acme-store-B.git
git remote update -p
git merge --allow-unrelated-histories  store-a/main
git merge --allow-unrelated-histories  store-b/main
find txns/
cd ..
``` 
In the meanwhile Store B has been really busy
```bash
# inside 'tackler-ng/tmp/acme-inc' 
cd acme-store-B
cp txns/store-b/txns-001.txn txns/store-b/txns-002.txn
cp txns/store-b/txns-001.txn txns/store-b/txns-003.txn
git add .
git commit -m 'store b'
git push
cd ..
```

Time to report numbers of holiday sales push
```bash
# inside 'tackler-ng/tmp/acme-inc' 
cd acme-hq
git remote update -p
git merge --allow-unrelated-histories  store-b/main
cd ..
```
Acme Inc HQ's reporting definition, this makes it possible to switch between working copy and repository with  `--input.storage "fs"`
```toml
[kernel]
strict = false
audit = { mode = false, hash = "SHA-256" }
timestamp = { default-time = 00:00:00, timezone = { name = "UTC" } }

[kernel.input]
storage = "git"
fs = { dir = "acme-hq/txns", suffix = "txn" }
git = { repository = "acme-hq/.git", dir = "txns", ref = "main", suffix = "txn" }

[transaction]
accounts    = { path = "none" }
commodities = { path = "none" }
tags        = { path = "none" }

### Report Configuration
[report]
report-timezone = "UTC"
scale = { min = 2, max = 4 }
accounts = [ "^Assets:", "^Warehouse:" ]
targets = [ "balance", "register" ]

balance       = { title = "Balance Report" }
balance-group = { title = "Balance Group Report", group-by = "month" }
register      = { title = "Register Report" }

### Export Configuration
[export]
targets = [ ]
equity = { equity-account = "Equity:Balance" }
```

```bash
# inside 'tackler-ng/tmp/acme-inc' 
../../target/release/tackler --config acme.toml
```

```
Git Storage
         commit : d0801cf2e7d2117cd04dd1c715450632514efe8f
      reference : main
      directory : txns
         suffix : .txn
        message : Merge remote-tracking branch 'store-b/main'
**********************************************************************************

Balance Report
--------------
                -4.00          -4.00 Unit  Warehouse:Acmerator
               168.00         168.00 €     Assets:Cashier
==========================
                -4.00 Unit
               168.00 €
##################################################################################
**********************************************************************************
Report Time Zone
        TZ name : UTC


Register Report
---------------
2024-12-17 'store a txn
            Warehouse:Acmerator                           -1.00              -1.00 Unit
            Assets:Cashier                                42.00              42.00 €
---------------------------------------------------------------------------------------
2024-12-17 'store b txn
            Warehouse:Acmerator                           -1.00              -2.00 Unit
            Assets:Cashier                                42.00              84.00 €
---------------------------------------------------------------------------------------
2024-12-17 'store b txn
            Warehouse:Acmerator                           -1.00              -3.00 Unit
            Assets:Cashier                                42.00             126.00 €
---------------------------------------------------------------------------------------
2024-12-17 'store b txn
            Warehouse:Acmerator                           -1.00              -4.00 Unit
            Assets:Cashier                                42.00             168.00 €
---------------------------------------------------------------------------------------
##################################################################################
```

And that's the beauty of these simple concepts: Each transaction is valid as single unit, so stores can run their own reports, and at the same time HQ can run reports over all transactions.  And git (and Gix!) makes all of this possible!

P.S. For this example all stores use the same Chart of Accounts, but it doesn't have to be so, everything would work similarly, even if stores would use e.g. own sub-accounts for warehouse balances.

**Reply by Byron** [reply]
This is great!

Git is used to bring together the independent views of each store, and if these stores use the same account names, HQ would be able to aggregate them naturally. Of course, nothing would prevent them from homing in on individual stores to get their view, or from keeping all account names separate, but aggregating sub-accounts or what not (I really have no clue what I am talking about 😅).

Now I am even more excited!

**Reply by 35VLG84** [reply]
Here is balance report so that the account structure is 
`Warehouse:[Store-A|Store-B]:Acmerator`

And these account structures could be enforced store by store basis, as there is that option for strict account name checking with the config (not activated in the example).

```
../../target/release/tackler --config acme.toml --reports balance
```

Here we have recursive balance view of that Warehouse account tree. The left most value is sum of the leaf account alone, the right most value is  sum of that account and all of its children.

So from here we can see that Warehouse at Store B is down -3 units and Warehouse at Store A is -1 Units, total 4 units sold.

```
Git Storage
         commit : fddc4528fb522f0e4c89d3b5e84abc3d5258ac32
      reference : main
      directory : txns
         suffix : .txn
        message : Merge remote-tracking branch 'store-a/main'

**********************************************************************************

Balance Report
--------------
                 0.00          -1.00 Unit  Warehouse:Store-A
                -1.00          -1.00 Unit  Warehouse:Store-A:Acmerator
                 0.00          -3.00 Unit  Warehouse:Store-B
                -3.00          -3.00 Unit  Warehouse:Store-B:Acmerator
               168.00         168.00 €     Assets:Cashier
==========================
                -4.00 Unit
               168.00 €
##################################################################################
```
P.S. Mandatory Git related thing: At some point JGit people were planning to have Raft based multi master backend for JGit (Git Ketch): https://projects.eclipse.org/projects/technology.jgit/releases/4.3/review 
That would be pretty cool for Acme Inc High-Available, fault tolerant register system ;)

(Disclaimer: Tackler is my hobby project and I'm not commercially building any such system, but even then, multi master Ketch based backend would be really cool)

**Comment by 35VLG84** [comment]
This is related to idea on  [Gitoxide Readme: Spin-Offs](https://github.com/GitoxideLabs/gitoxide?tab=readme-ov-file#ideas-for-spin-offs): Time Tracker

This is real time tracker entry for Tackler-NG project: the `cid: ... ` is commit id for this commit on [tackler-ng: bfc8f688536da27](https://github.com/e257-fi/tackler-ng/commit/bfc8f688536da27) 

```
$ cat journal/txns/2023/W16/7/20230423T181100-d76801bb-b0b0-48e8-a191-fad1ed934721.txn
2023-04-23T18:11:00 'cid: bfc8f688536da27
   # uuid: d76801bb-b0b0-48e8-a191-fad1ed934721
   Source:Time:Stopwatch:Start    1682254352 s @ 0.00027778 h ; 2023-04-23T15:52:32.0644337+03:00
   Source:Time:Stopwatch:Stop    -1682262660 s @ 0.00027778 h ; 2023-04-23T18:11:00.356301+03:00
   Timelog:Projects:Tackler:TNG
```
It's using seconds as based unit for time tracking and converting those to hours with ratio of 0.00027778. The entry is generated with separate tool. The tool is committing directly to Git repository, but it's currently Scala+JGIt. I'm planning to port it to Rust+Gix eventually.

**Reply by Byron** [reply]
Thanks so much for pointing this out!

Now that I learned about tackler I definitely thought about integrating with it with such a timetracking tool, but I also realise that it's very possible and just as reasonable to do time-tracking directly in tackler.

I am definitely looking forward to the port!

Independently of that, I recommend also taking a look at [`jiff`](https://crates.io/search?q=jiff) to replace `time` - it's really good and by a maintainer I trust more than myself :D.

**Reply by 35VLG84** [reply]
Thanks, that's nice tip, as good datetime library with included support for TZ would be really nice.
